crates/virtual-node/src/lib.rs

```diff
@@ -101,54 +108,99 @@ impl VirtualNode {
     /// ```
-    pub fn text(text: &str) -> VirtualNode {
-        VirtualNode {
-            tag: "".to_string(),
-            props: HashMap::new(),
-            events: Events(HashMap::new()),
-            children: Some(vec![]),
-            text: Some(text.to_string()),
-        }
+    pub fn text(text: &str) -> Self {
+        VirtualNode::Text(text.into())
     }
-}
 
-/// A web_sys::Element along with all of the closures that were created for that element's
-/// events and all of it's child element's events.
-pub struct CreatedElement {
-    /// An Element that was created from a VirtualNode
-    pub element: Element,
-    /// A map of an element's unique identifier along with all of the Closures for that element.
+    /// Create a new `VirtualDomElement`.
     ///
-    /// The DomUpdater uses this to look up elements and see if they're still in the page. If not
-    /// the refernce that we maintain to their closure will be dropped, thus freeing the Closure's
-    /// memory.
-    pub closures: HashMap<u32, Vec<DynClosure>>,
-}
+    /// This is the variant that is contained inside `VirtualNode::Element`. It
+    /// can be turned into a `VirtualNode` using the from/into.
+    pub fn element_variant<S: Into<String>>(tag: S) -> VirtualNodeElement {
+        VirtualNodeElement::new(tag)
+    }
 
-impl Deref for CreatedElement {
-    type Target = Element;
+    /// Create a new `VirtualDomText`.
+    ///
+    /// This is the variant that is contained inside `VirtualNode::Text`. It
+    /// can be turned into a `VirtualNode` using the from/into.
+    pub fn text_variant<S: Into<String>>(text: S) -> VirtualNodeText {
+        VirtualNodeText { text: text.into() }
+    }
 
-    fn deref(&self) -> &Self::Target {
-        &self.element
+    /// Cast into a `VirtualNodeElement` reference, if this is an `Element` variant.
+    pub fn as_element_variant_ref(&self) -> Option<&VirtualNodeElement> {
+        match self {
+            VirtualNode::Element(ref element_node) => Some(element_node),
+            _ => None,
+        }
     }
-}
 
-fn create_unique_identifier() -> u32 {
-    let mut elem_unique_id = ELEM_UNIQUE_ID.lock().unwrap();
+    /// Cast into a mutable `VirtualNodeElement` reference, if this is an `Element` variant.
+    pub fn as_element_variant_ref_mut(&mut self) -> Option<&mut VirtualNodeElement> {
+        match self {
+            VirtualNode::Element(ref mut element_node) => Some(element_node),
+            _ => None,
+        }
+    }
 
-    *elem_unique_id += 1;
+    /// Cast into a `VirtualNodeText` reference, if this is a `Text` variant.
+    pub fn as_text_variant_ref(&self) -> Option<&VirtualNodeText> {
+        match self {
+            VirtualNode::Text(ref text_node) => Some(text_node),
+            _ => None,
+        }
+    }
 
-    *elem_unique_id
+    /// Cast into a mutable `VirtualNodeText` reference, if this is a `Text` variant.
+    pub fn as_text_variant_ref_mut(&mut self) -> Option<&mut VirtualNodeText> {
```

Comment:
I think that `as` vs. `into` conveys borrow vs reference so the word `ref` isn't needed in these methods.

Reply:
Yep, you're right, I just noticed that too. The ref is redundant :slightly_smiling_face: